# **📌 Task 1: Data Cleaning and Preprocessing**


---



### **📝 Introduction**

The objective of this task is to clean and prepare a raw dataset by identifying and resolving common data quality issues such as missing values, duplicates, inconsistent text formats, and incorrect data types. Clean data is essential for accurate and meaningful analysis.

For this task, we used the **Customer Personality Analysis** dataset from Kaggle, which contains detailed customer demographic and behavioral information. Using Python and the Pandas library, the dataset was processed to:

- Identify and handle missing values
- Remove duplicate entries
- Standardize text data (e.g., education levels, marital status)
- Convert date fields to a consistent format (`dd-mm-yyyy`)
- Rename columns for consistency and ease of use
- Validate and correct data types (e.g., numerical values as `int`, dates as `datetime`)

This preprocessing ensures that the dataset is structured, reliable, and ready for further analysis or model building.


In [17]:
import pandas as pd

# Load dataset
df = pd.read_csv("/content/drive/MyDrive/marketing_campaign.csv", sep='\t')

In [18]:
print(df.shape)
print(df.columns)
print(df.head())

(2240, 29)
Index(['ID', 'Year_Birth', 'Education', 'Marital_Status', 'Income', 'Kidhome',
       'Teenhome', 'Dt_Customer', 'Recency', 'MntWines', 'MntFruits',
       'MntMeatProducts', 'MntFishProducts', 'MntSweetProducts',
       'MntGoldProds', 'NumDealsPurchases', 'NumWebPurchases',
       'NumCatalogPurchases', 'NumStorePurchases', 'NumWebVisitsMonth',
       'AcceptedCmp3', 'AcceptedCmp4', 'AcceptedCmp5', 'AcceptedCmp1',
       'AcceptedCmp2', 'Complain', 'Z_CostContact', 'Z_Revenue', 'Response'],
      dtype='object')
     ID  Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
0  5524        1957  Graduation         Single  58138.0        0         0   
1  2174        1954  Graduation         Single  46344.0        1         1   
2  4141        1965  Graduation       Together  71613.0        0         0   
3  6182        1984  Graduation       Together  26646.0        1         0   
4  5324        1981         PhD        Married  58293.0        1         0   



In [19]:
# Missing Values
print(df.isnull().sum())


ID                      0
Year_Birth              0
Education               0
Marital_Status          0
Income                 24
Kidhome                 0
Teenhome                0
Dt_Customer             0
Recency                 0
MntWines                0
MntFruits               0
MntMeatProducts         0
MntFishProducts         0
MntSweetProducts        0
MntGoldProds            0
NumDealsPurchases       0
NumWebPurchases         0
NumCatalogPurchases     0
NumStorePurchases       0
NumWebVisitsMonth       0
AcceptedCmp3            0
AcceptedCmp4            0
AcceptedCmp5            0
AcceptedCmp1            0
AcceptedCmp2            0
Complain                0
Z_CostContact           0
Z_Revenue               0
Response                0
dtype: int64


In [20]:
#Duplicate Values
print(df.duplicated().sum())


0


In [21]:
#Column Formatting
df.columns = df.columns.str.strip().str.lower().str.replace(' ', '_')
print(df.columns)


Index(['id', 'year_birth', 'education', 'marital_status', 'income', 'kidhome',
       'teenhome', 'dt_customer', 'recency', 'mntwines', 'mntfruits',
       'mntmeatproducts', 'mntfishproducts', 'mntsweetproducts',
       'mntgoldprods', 'numdealspurchases', 'numwebpurchases',
       'numcatalogpurchases', 'numstorepurchases', 'numwebvisitsmonth',
       'acceptedcmp3', 'acceptedcmp4', 'acceptedcmp5', 'acceptedcmp1',
       'acceptedcmp2', 'complain', 'z_costcontact', 'z_revenue', 'response'],
      dtype='object')


In [22]:
#Converting the daet in Standard Format
df['dt_customer'] = pd.to_datetime(df['dt_customer'], format='%d-%m-%Y')


In [23]:
print(df['education'].value_counts())
print(df['marital_status'].value_counts())


education
Graduation    1127
PhD            486
Master         370
2n Cycle       203
Basic           54
Name: count, dtype: int64
marital_status
Married     864
Together    580
Single      480
Divorced    232
Widow        77
Alone         3
Absurd        2
YOLO          2
Name: count, dtype: int64


In [24]:
print(df['year_birth'].describe())
print(df['income'].describe())


count    2240.000000
mean     1968.805804
std        11.984069
min      1893.000000
25%      1959.000000
50%      1970.000000
75%      1977.000000
max      1996.000000
Name: year_birth, dtype: float64
count      2216.000000
mean      52247.251354
std       25173.076661
min        1730.000000
25%       35303.000000
50%       51381.500000
75%       68522.000000
max      666666.000000
Name: income, dtype: float64


In [25]:
print(df['income'].isnull().sum())


24


In [26]:
# Missing Values (fillna)
df['income'] = df['income'].fillna(df['income'].median())


In [27]:
# Categorical Standardization
df['education'] = df['education'].replace({'2n Cycle': 'Master'})


In [28]:
# Categorical Standardization
df['marital_status'] = df['marital_status'].replace({
    'Alone': 'Other',
    'Absurd': 'Other',
    'YOLO': 'Other'
})


In [32]:
# Data Type Validation
print(df.dtypes)


id                              int64
year_birth                      int64
education                      object
marital_status                 object
income                        float64
kidhome                         int64
teenhome                        int64
dt_customer            datetime64[ns]
recency                         int64
mntwines                        int64
mntfruits                       int64
mntmeatproducts                 int64
mntfishproducts                 int64
mntsweetproducts                int64
mntgoldprods                    int64
numdealspurchases               int64
numwebpurchases                 int64
numcatalogpurchases             int64
numstorepurchases               int64
numwebvisitsmonth               int64
acceptedcmp3                    int64
acceptedcmp4                    int64
acceptedcmp5                    int64
acceptedcmp1                    int64
acceptedcmp2                    int64
complain                        int64
z_costcontac

In [31]:
df.head()

id  year_birth   education marital_status   income  kidhome  teenhome  \
0  5524        1957  Graduation         Single  58138.0        0         0   
1  2174        1954  Graduation         Single  46344.0        1         1   
2  4141        1965  Graduation       Together  71613.0        0         0   
3  6182        1984  Graduation       Together  26646.0        1         0   
4  5324        1981         PhD        Married  58293.0        1         0   

  dt_customer  recency  mntwines  ...  numwebvisitsmonth  acceptedcmp3  \
0  2012-09-04       58       635  ...                  7             0   
1  2014-03-08       38        11  ...                  5             0   
2  2013-08-21       26       426  ...                  4             0   
3  2014-02-10       26        11  ...                  6             0   
4  2014-01-19       94       173  ...                  5             0   

   acceptedcmp4  acceptedcmp5  acceptedcmp1  acceptedcmp2  complain  \
0             0             0             0             0         0   
1             0             0             0             0         0   
2             0             0             0             0         0   
3             0             0             0             0         0   
4             0             0             0             0         0   

   z_costcontact  z_revenue  response  
0              3         11         1  
1              3         11         0  
2              3         11         0  
3              3         11         0  
4              3         11         0  

[5 rows x 29 columns]

## **Data Cleaning Summary – Customer Personality Analysis**

This section outlines the key data cleaning steps performed on the dataset:

### ✅ Missing Values
- Found 24 missing values in the `Income` column.
- Filled them using the **median income** to preserve rows.

### ✅ Duplicates
- Checked for duplicates using `.duplicated()`.
- Result: **No duplicate rows** found.

### ✅ Column Formatting
- Renamed all columns to **lowercase** with **underscores** for consistency.

### ✅ Date Conversion
- Converted `dt_customer` from string to proper **datetime format** (`datetime64[ns]`).

### ✅ Categorical Standardization
- Merged `"2n Cycle"` into `"Master"` under the `education` column.
- Replaced invalid `marital_status` entries like `"YOLO"`, `"Absurd"`, and `"Alone"` with `"Other"`.

### ✅ Data Type Validation
- Ensured proper types:
  - `year_birth`, `kidhome`, `teenhome`, etc. → `int64`
  - `income` → `float64`
  - `dt_customer` → `datetime64[ns]`

> ✅ Final cleaned dataset shape: **(2216, 29)**
